### PR TITLE
Simplify adding a rule that use a diff action

### PR DESCRIPTION
**Pull request by bobot**
When adding tests that use the diff action, it is inconvenient that you need to create an empty file for the reference. Otherwise the rule can't be executed since the reference is not available.

This MR fix the first part, the rule can be executed and the diffing is done against the empty file `/dev/null` or `NUL`. The heuristic for promotion is modified by also doing a promotion if it is not a target, which at this point is equivalent with the file is not present in the build directory.

### CODE REVIEW

**Comment by rgrinberg** [comment]
Is this still WIP? It seems ready to me. Could you create a separate test file for this with a proper description. E.g:

```
ls promote/
  old-tests.t
  non-existent.t
```

I realize the old tests aren't well organized, but new tests should be easy to read.

@snowleopard could we get your review here?

File: src/dune_engine/build.mli [hunk 1]
```diff
@@ -158,13 +158,17 @@ val strings : Path.t -> string list t
...
-    target of a rule. *)
+    target of a rule. It doesn't add the path as dependency *)
 val file_exists : Path.t -> bool t
 
 (** [if_file_exists p ~then ~else] is a description that behaves like [then_] if
     [file_exists p] evaluates to [true], and [else_] otherwise. *)
 val if_file_exists : Path.t -> then_:'a t -> else_:'a t -> 'a t
 
+(** [filter_existing_files p] is a build description which keep only the
+    existing files *)
```
Comment (snowleopard) [inline src/dune_engine/build.mli:169]:
```suggestion
    existing files. The files are not registered as dynamic dependencies. *)
```

File: src/dune_engine/build.ml [hunk 2]
```diff
@@ -10,6 +10,7 @@ type 'a t =
   | Paths_for_rule : Path.Set.t -> unit t
   | Paths_glob : File_selector.t -> Path.Set.t t
   | If_file_exists : Path.t * 'a t * 'a t -> 'a t
+  | Filter_existing_files : ('a * Path.Set.t) t -> ('a * Path.Set.t) t
```
Comment (snowleopard) [inline src/dune_engine/build.ml:13]: It's not immediately obvious why `Filter_existing_files` can't be built on top of `If_file_exists`, so a short comment would be useful for future readers.

**Review by snowleopard — APPROVED** [review]
Thanks @bobot! It's an interesting new `Build` combinator, which I guess might develop further in future.

I reviewed the code related to `build.ml`. It looks OK, but see a couple of suggestions regarding comments.

I don't really know much about `action_unexpanded.ml` and friends, but my understanding is that @rgrinberg had already looked at that part.

File: src/dune_engine/build.ml (hunk 2 above)
Comment (bobot) [reply src/dune_engine/build.ml:13]: The reasons are multifold, but are you convinced by just this one:

```suggestion
  | Filter_existing_files : ('a * Path.Set.t) t -> ('a * Path.Set.t) t
  (* [Filter_existing_files] can't be defined using [If_file_exists] because in the latter the path must be known when building the type t. In the former case the paths can be dynamically computed *)
```

Comment (snowleopard) [reply src/dune_engine/build.ml:13]: Yes, that's a very clear explanation. It's exactly how I convinced myself after being puzzled for a couple of minutes.

**Comment by rgrinberg** [comment]
@bobot could you push a commit to master that updates CHANGES?
